sor/src/main/java/com/bazaarvoice/emodb/sor/db/astyanax/DAOModule.java

```diff
@@ -18,27 +23,122 @@
+        bind(DataWriterDAO.class).annotatedWith(AstyanaxWriterDAODelegate.class).to(CqlDataWriterDAO.class).asEagerSingleton();
+        bind(DataWriterDAO.class).to(CqlDataWriterDAO.class).asEagerSingleton();
         bind(DataPurgeDAO.class).to(AstyanaxDataWriterDAO.class).asEagerSingleton();
+        bind(MigratorWriterDAO.class).to(CqlDataWriterDAO.class).asEagerSingleton();
+        bind(MigratorReaderDAO.class).to(CqlDataReaderDAO.class).asEagerSingleton();
 
         // Explicit bindings so objects don't get created as a just-in-time binding in the root injector.
         // This needs to be done for just about anything that has only public dependencies.
-        bind(AstyanaxDataReaderDAO.class).asEagerSingleton();
+        bind(AstyanaxDataWriterDAO.class).asEagerSingleton();
 
         expose(DataReaderDAO.class);
         expose(DataWriterDAO.class);
         expose(DataCopyDAO.class);
         expose(DataPurgeDAO.class);
+        expose(MigratorReaderDAO.class);
+        expose(MigratorWriterDAO.class);
     }
 
     @Provides
     @Singleton
     ChangeEncoder provideChangeEncoder(DataStoreConfiguration configuration) {
         return new DefaultChangeEncoder(configuration.getDeltaEncodingVersion());
     }
-}
+
+    @Provides
+    @WriteToOld
+    boolean provideWriteToOld(DataStoreConfiguration configuration) {
+        int migrationPhase = configuration.getMigrationPhase();
+        checkArgument(migrationPhase >= 0 && migrationPhase <= 3, "Invalid Migration Phase");
+        return migrationPhase == 0 || migrationPhase == 1 || migrationPhase == 2;
+    }
+
+    @Provides
+    @WriteToNew
+    boolean provideWriteToNew(DataStoreConfiguration configuration) {
+        int migrationPhase = configuration.getMigrationPhase();
+        checkArgument(migrationPhase >= 0 && migrationPhase <= 3, "Invalid Migration Phase");
+        return migrationPhase == 1 || migrationPhase == 2 || migrationPhase == 3;
+    }
+
+    @Provides
+    @Singleton
+    DataReaderDAO provideDataReaderDAO(DataStoreConfiguration configuration, @CqlReaderDAODelegate DataReaderDAO delegate,
+                                       PlacementCache placementCache, CqlDriverConfiguration driverConfig,
+                                       ChangeEncoder changeEncoder, MetricRegistry metricRegistry,
+                                       DAOUtils daoUtils, @PrefixLength int prefixLength) {
+
+        int migrationPhase = configuration.getMigrationPhase();
+        checkArgument(migrationPhase >= 0 && migrationPhase <= 3, "Invalid Migration Phase");
+
+        if (migrationPhase == 0 || migrationPhase == 1) {
+            return new CqlDataReaderDAO(delegate, placementCache, driverConfig, changeEncoder, metricRegistry);
+        }
+
+        return new CqlBlockedDataReaderDAO(delegate, placementCache, driverConfig, changeEncoder, metricRegistry, daoUtils, prefixLength);
```

Comment:
Although this works it's bypassing much of the benefits of dependency injection by requiring this module to know the details of how to construct the data readers.  If you add the following to `configure()`:

```
        bind(CqlDataReaderDAO.class).asEagerSingleton();
        bind(CqlBlockedDataReaderDAO.class).asEagerSingleton();
```

then you can use providers to create the the instances:

```
    DataReaderDAO provideDataReaderDAO(DataStoreConfiguration configuration,
                                       Provider<CqlDataReaderDAO> legacyReader,
                                       Provider<CqlBlockedDataReaderDAO> blockedReader) {
        ...

        if (migrationPhase == 0 || migrationPhase == 1) {
            return legacyReader.get();
        }

        return blockedReader.get();
    }
```

Reply:
done